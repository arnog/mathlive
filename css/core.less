// The mu (mathematical unit) is defined by TeX as 1/18 em
@mu: 1/18em;

@system: system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto',
  'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
  sans-serif;

  .ML__container {
    min-height: auto !important;
    --_hue: var(--hue, 212);
    --_placeholder-color: var(--placeholder-color, hsl(var(--_hue), 40%, 49%));
    --_placeholder-opacity: var(--placeholder-opacity, 0.4);
    --_text-font-family: var(--text-font-family, @system);
  }
  

.ML__sr-only {
  position: absolute;
  width: 1px;
  height: 1px;
  margin: -1px;
  padding: 0;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  clip-path: inset(50%);
  white-space: nowrap;
  border: 0;
}

.ML__is-inline {
  display: inline-block;
}

.ML__base {
  visibility: inherit;
  display: inline-block;
  position: relative;
  cursor: text;
  padding: 0;
  margin: 0;
  box-sizing: content-box;
  border: 0;
  outline: 0;
  vertical-align: baseline;
  font-weight: inherit;
  font-family: inherit;
  font-style: inherit;
  text-decoration: none;
  // Fix width of containers of negative spaces, working around Chrome bug.
  width: min-content;
}

.ML__strut,
.ML__strut--bottom {
  display: inline-block;
  min-height: 0.5em;
}

.ML__small-delim {
  font-family: KaTeX_Main;
}

/* Text mode */
.ML__text {
  font-family: var(--_text-font-family);
  white-space: pre;
}

/* Use cmr for 'math upright' */
.ML__cmr {
  font-family: KaTeX_Main;
  font-style: normal;
}

.ML__mathit {
  font-family: KaTeX_Math;
  /* The KaTeX_Math font is italic by default, so the font-style below is only 
     useful when a fallback font is used
  */
  font-style: italic;
}

.ML__mathbf {
  font-family: KaTeX_Main; // Use upright font
  font-weight: bold;
}

/* Lowercase greek symbols should stick to math font when \mathbf is applied 
   to match TeX idiosyncratic behavior */
.lcGreek.ML__mathbf {
  font-family: KaTeX_Math;
  font-weight: normal;
}

.ML__mathbfit {
  font-family: KaTeX_Math;
  font-weight: bold;
  font-style: italic;
}

.ML__ams {
  font-family: KaTeX_AMS;
}

/* Blackboard */
.ML__bb {
  font-family: KaTeX_AMS;
}

.ML__cal {
  font-family: KaTeX_Caligraphic;
}

.ML__frak {
  font-family: KaTeX_Fraktur;
}

.ML__tt {
  font-family: KaTeX_Typewriter;
}

.ML__script {
  font-family: KaTeX_Script;
}

.ML__sans {
  font-family: KaTeX_SansSerif;
}
.ML__series_ul {
  font-weight: 100;
}
.ML__series_el {
  font-weight: 100;
}
.ML__series_l {
  font-weight: 200;
}
.ML__series_sl {
  font-weight: 300;
}

.ML__series_sb {
  font-weight: 500;
}
.ML__bold,
.ML__boldsymbol {
  font-weight: 700;
}
.ML__series_eb {
  font-weight: 800;
}
.ML__series_ub {
  font-weight: 900;
}

.ML__series_uc {
  font-stretch: ultra-condensed;
}
.ML__series_ec {
  font-stretch: extra-condensed;
}
.ML__series_c {
  font-stretch: condensed;
}
.ML__series_sc {
  font-stretch: semi-condensed;
}
.ML__series_sx {
  font-stretch: semi-expanded;
}
.ML__series_x {
  font-stretch: expanded;
}
.ML__series_ex {
  font-stretch: extra-expanded;
}
.ML__series_ux {
  font-stretch: ultra-expanded;
}

.ML__it {
  font-style: italic;
}
.ML__shape_ol {
  -webkit-text-stroke: 1px black;
  text-stroke: 1px black;
  color: transparent;
}
.ML__shape_sc {
  font-variant: small-caps;
}
.ML__shape_sl {
  font-style: oblique;
}

/* First level emphasis */
.ML__emph {
  color: #bc2612;
}
/* Second level emphasis */
.ML__emph .ML__emph {
  color: #0c7f99;
}

// Style used to highlight a portion of the expression when speaking
// it with synchronized highlighting.
// See https://www.w3.org/WAI/RD/2012/text-customization/r11
// and https://www.cs.cmu.edu/~jbigham/pubs/pdfs/2017/colors.pdf
.ML__highlight {
  color: #007cb2; // Complementary to peach
  background: #edd1b0; // Peach
}

.ML__center {
  text-align: center;
}

.ML__left {
  text-align: left;
}

.ML__right {
  text-align: right;
}


.ML__label_padding {
  padding: 0 0.5em;
}

.ML__frac-line {
  width: 100%;
  min-height: 1px;

  &:after {
    content: '';
    display: block;
    margin-top: ~"max(-1px, -0.04em)";
    min-height: ~"max(1px, 0.04em)";
    /* Ensure the line is visible when printing even if "turn off background images" is on*/
    -webkit-print-color-adjust: exact;
    print-color-adjust: exact;
    /* There's a bug since Chrome 62 where 
      sub-pixel border lines don't draw at some zoom 
      levels (110%, 90%). 
      Setting the min-height used to work around it, but that workaround
      broke in Chrome 84 or so.
      Setting the background (and the min-height) seems to work for now.
      */
    background: currentColor;
    box-sizing: content-box; /* Vuetify sets the box-sizing to inherit 
            causes the fraction line to not draw at all sizes (see #26) */
    /* On some versions of Firefox on Windows, the line fails to 
            draw at some zoom levels, but setting the transform triggers
            the hardware accelerated path, which works */
    transform: translate(0, 0);
  }
}

.ML__sqrt {
  display: inline-block;
}

.ML__sqrt-sign {
  display: inline-block;
  position: relative;
}

.ML__sqrt-line {
  display: inline-block;
  height: ~"max(1px, 0.04em)";
  width: 100%;

  &:before {
    content: '';
    display: block;
    margin-top: ~"min(-1px, -0.04em)";
    // margin-top: -1px;
    min-height: ~"max(1px, 0.04em)";
    /* Ensure the line is visible when printing even if "turn off background images" is on*/
    -webkit-print-color-adjust: exact;
    print-color-adjust: exact;
    background: currentColor;
    /* On some versions of Firefox on Windows, the line fails to 
            draw at some zoom levels, but setting the transform triggers
            the hardware accelerated path, which works */
    transform: translate(0, 0);
  }

  &:after {
    border-bottom-width: 1px;
    content: ' ';
    display: block;
    margin-top: -0.1em;
  }
}

.ML__sqrt-index {
  // These values are taken from the definition of `\r@@t`,
  // `\mkern 5mu` and `\mkern -10mu`.
  margin-left: 5 * (@mu);
  margin-right: -10 * (@mu);
}

.ML__delim-size1 {
  font-family: KaTeX_Size1;
}
.ML__delim-size2 {
  font-family: KaTeX_Size2;
}
.ML__delim-size3 {
  font-family: KaTeX_Size3;
}
.ML__delim-size4 {
  font-family: KaTeX_Size4;
}

.ML__delim-mult {
  .delim-size1 > span {
    font-family: KaTeX_Size1;
  }
  .delim-size4 > span {
    font-family: KaTeX_Size4;
  }
}

.ML__accent-body > span {
  font-family: KaTeX_Main;
  width: 0;
}

.ML__accent-vec {
  position: relative;
  left: 0.24em;
}

/** The markup for a LaTeX formula, either in an editable mathfield or 
    in a static display.
*/
.ML__latex {
  @ptperem: 10; // This value should equal PT_PER_EM om `font-metrics.ts`

  @thinspace: 0.16667em; // 3mu
  @mediumspace: 0.22222em; // 4mu
  @thickspace: 0.27778em; // 5mu

  display: inline-block;

  direction: ltr;

  text-align: left;
  text-indent: 0;
  // Prevent a rendering bug that misplaces \vec in Chrome.
  text-rendering: auto;
  font-family: KaTeX_Main, 'Times New Roman', serif;
  font-style: normal;
  font-size-adjust: none;
  font-stretch: normal;
  font-variant-caps: normal;
  letter-spacing: normal;

  // This value is critical in getting the vertical alignment correct.
  // Check with 1+\frac{5}{\frac{3}{\frac{7}{29}}}+\left(\frac{5}{\frac{3}{\frac{7}{29}}}\right)
  line-height: 1.2;

  word-wrap: normal;
  word-spacing: normal;
  white-space: nowrap;

  text-shadow: none;
  -webkit-user-select: none;
  user-select: none;

  width: min-content;

  .style-wrap {
    position: relative;
  }

  .ML__mfrac {
    display: inline-block;
  }
  
  .ML__left-right {
    display: inline-block;
  }

  .ML__vlist-t {
    display: inline-table;
    table-layout: fixed;
    border-collapse: collapse;
  }

  .ML__vlist-r {
    display: table-row;
  }

  .ML__vlist {
    display: table-cell;    // Inside a table-cell, vertical-align aligns the content of the cell, not the cell itself
    vertical-align: bottom;
    position: relative;

    > span {
      display: block;
      height: 0;
      position: relative;

      > span {
        display: inline-block;
      }
      > .ML__pstrut {
        overflow: hidden;
        width: 0;
      }
    }
  }

  .ML__vlist-t2 {
    margin-right: -2px;
  }

  .ML__vlist-s {
    // This cell solves Safari rendering problems. It has text content, so
    // its baseline is used for the table. A very small font avoids line-box
    // issues; absolute units prevent user font-size overrides from breaking
    // rendering. Safari refuses to make the box zero-width, so we give it
    // a known width and compensate with negative right margin on the
    // inline-table. To prevent the "width: min-content" Chrome workaround
    // from shrinking this box, we also set min-width.
    display: table-cell;
    vertical-align: bottom;
    font-size: 1px;
    width: 2px;
    min-width: 2px;
  }
  .ML__msubsup {
    text-align: left;
  }

  .ML__negativethinspace {
    display: inline-block;
    margin-left: -@thinspace;
    height: 0.71em;
  }

  .ML__thinspace {
    display: inline-block;
    width: @thinspace;
    height: 0.71em;
  }

  .ML__mediumspace {
    display: inline-block;
    width: @mediumspace;
    height: 0.71em;
  }

  .ML__thickspace {
    display: inline-block;
    width: @thickspace;
    height: 0.71em;
  }

  .ML__enspace {
    display: inline-block;
    width: 0.5em;
    height: 0.71em;
  }

  .ML__quad {
    display: inline-block;
    width: 1em;
    height: 0.71em;
  }

  .ML__qquad {
    display: inline-block;
    width: 2em;
    height: 0.71em;
  }

  .ML__llap,
  .ML__rlap {
    width: 0;
    position: relative;
    display: inline-block;

    > .ML__inner {
      position: absolute;
    }

    > .ML__fix {
      display: inline-block;
    }
  }

  .ML__llap > .ML__inner {
    right: 0;
  }

  .ML__rlap > .ML__inner {
    left: 0;
  }

  .ML__rule {
    display: inline-block;
    border: solid 0;
    position: relative;
    box-sizing: border-box;
  }

  .overline .overline-line,
  .underline .underline-line {
    width: 100%;

    &:before {
      content: '';
      border-bottom-style: solid;
      border-bottom-width: ~"max(1px, 0.04em)";
      -webkit-print-color-adjust: exact;
      print-color-adjust: exact;
      display: block;
    }

    &:after {
      border-bottom-style: solid;
      // Since Chrome 62 there are issues rendering sub-pixel borders at 
      // certain zoom levels (90%, 110%). Forcing it to at least one pixels 
      // seems to work around the issue.
      border-bottom-width: ~"max(1px, 0.04em)";
      -webkit-print-color-adjust: exact;
      print-color-adjust: exact;
        content: '';
      display: block;
      margin-top: -1px;
    }
  }

  .ML__stretchy {
    display: block;
    position: absolute;
    width: 100%;
    left: 0;
    overflow: hidden;
    &:before,
    &:after {
      content: '';
    }
    svg {
      display: block;
      position: absolute;
      width: 100%;
      height: inherit;
      fill: currentColor;
      stroke: currentColor;
      fill-rule: nonzero;
      fill-opacity: 1;
      stroke-width: 1;
      stroke-linecap: butt;
      stroke-linejoin: miter;
      stroke-miterlimit: 4;
      stroke-dasharray: none;
      stroke-dashoffset: 0;
      stroke-opacity: 1;
    }
  }

  .slice-1-of-2 {
    display: inline-flex;
    position: absolute;
    left: 0;
    width: 50.2%;
    overflow: hidden;
  }
  .slice-2-of-2 {
    display: inline-flex;
    position: absolute;
    right: 0;
    width: 50.2%;
    overflow: hidden;
  }
  .slice-1-of-3 {
    display: inline-flex;
    position: absolute;
    left: 0;
    width: 25.1%;
    overflow: hidden;
  }
  .slice-2-of-3 {
    display: inline-flex;
    position: absolute;
    left: 25%;
    width: 50%;
    overflow: hidden;
  }
  .slice-3-of-3 {
    display: inline-flex;
    position: absolute;
    right: 0;
    width: 25.1%;
    overflow: hidden;
  }
  .slice-1-of-1 {
    display: inline-flex;
    position: absolute;
    width: 100%;
    left: 0;
    overflow: hidden;
  }

  .ML__nulldelimiter {
    display: inline-block;
  }

  .ML__op-group {
    display: inline-block;
  }

  .ML__op-symbol {
    position: relative;

    &.ML__small-op {
      font-family: KaTeX_Size1;
    }
    &.ML__large-op {
      font-family: KaTeX_Size2;
    }
  }

  .ML__mtable {
    .ML__vertical-separator {
      display: inline-block;
      // margin and border-right are set in Javascript
      min-width: 1px; // Prevent Chrome from omitting a line.
      box-sizing: border-box;
    }

    .ML__arraycolsep {
      display: inline-block;
    }

    .col-align-m > .ML__vlist-t {
      text-align: center;
    }

    .col-align-c > .ML__vlist-t {
      text-align: center;
    }

    .col-align-l > .ML__vlist-t {
      text-align: left;
    }

    .col-align-r > .ML__vlist-t {
      text-align: right;
    }
  }
}

// Wavy pattern
.ML__error {
  display: inline-block;
  background-image: radial-gradient(
    ellipse at center,
    hsl(341, 100%, 40%),
    rgba(0, 0, 0, 0) 70%
  );
  background-color: hsla(341, 100%, 40%, 10%);

  background-repeat: repeat-x;
  background-size: 3px 3px;
  padding-bottom: 3px;
  background-position: 0 100%;
}

.ML__error > .ML__error {
  background: transparent;
  padding: 0;
}


.ML__placeholder {
  color:  var(--_placeholder-color);
  opacity: var(--_placeholder-opacity);
  padding-left: 0.4ex;
  padding-right: 0.4ex;
  font-family: @system;
}

.ML__notation {
  position: absolute;
  box-sizing: border-box;
  line-height: 0;
}


/* This class is used to implement the `\mathtip` and `\texttip` commands
   For UI elements, see `[data-ML__tooltip]`
*/
.ML__tooltip-container {
  position: relative;
  transform: scale(0); // To override the overflow:hidden

  .ML__tooltip-content {
    position: fixed;
    display: inline-table;
    visibility: hidden;
    z-index: 2;
    width: max-content;
    max-width: 400px;
    padding: 12px 12px;
    border-radius: 8px;

    background: #616161; // Grey 700
    --_selection-color: #fff;
    color: #fff;
    box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14), 0 1px 5px 0 rgba(0, 0, 0, 0.12),
      0 3px 1px -2px rgba(0, 0, 0, 0.2);
    opacity: 0;
    // animation: .150s tooltipFadeOut cubic-bezier(0.4, 0.0, 1, 1) forwards;
    transition: opacity 0.15s cubic-bezier(0.4, 0, 1, 1);

    .ML__text {
      white-space: normal; // Make text wrap
    }
    .ML__base {
      display: contents;
    }
  }
  &:hover .ML__tooltip-content {
    visibility: visible; // Visible triggers the animation...
    opacity: 1; // ... but we start fully transparent to
    // give ourselves a delay before showing
    font-size: 0.75em;
    transform: scale(1) translate(0, 3em);
  }
}

