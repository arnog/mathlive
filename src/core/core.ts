import './atom-class';
import './atom';
import './context';
import './delimiters';
import './mathstyle';
import './parser';
import './box';
import './font-metrics';
import './modes';

export * from './atom-class';
export * from './atom';
export * from './context';
export * from './delimiters';
export * from './mathstyle';
export * from './parser';
export * from './box';
export * from './font-metrics';
