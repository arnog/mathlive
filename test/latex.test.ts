import {
  convertLatexToMarkup,
  validateLatex,
} from '../src/public/mathlive-ssr';
import { parseLatex } from '../src/core/parser';
import { Atom } from '../src/core/atom-class';

function markupAndError(formula: string): [string, string] {
  const markup = convertLatexToMarkup(formula, { defaultMode: 'math' });
  const errors = validateLatex(formula);
  if (errors.length === 0) return [markup, 'no-error'];
  return [markup, errors[0].code];
}

function error(expression: string) {
  const errors = validateLatex(expression);
  if (errors.length === 0) return 'no-error';
  return errors[0].code;
}

describe('BASIC PARSING', () => {
  test.each([
    'x',
    ' x ', // Spaces do not matter
    '%', // '%' is start of comment
    '% comment',
    'x % comment',
    'x',
    '-12',
    '1234|/@.`abcdefgzABCDEFGZ', // Basic literals
    'a b', // Spaces are ignored
    'ab', // Same as previous
    'a~b', // ~ is space, same as previous
    'a\\space b',
    '{a}b', // Group
    '{-}', // Operator in group
    '-a', // Spacing as unary operator
    'a-', // Spacing as postfix operator
    'a-b', // Spacing as infix operator
    'a\nb',
    'a=1}',
    'a=1{', // Syntax error
    'a=1{}', // Valid
  ])('%#/ %p renders correctly', (x) => {
    expect(markupAndError(x)).toMatchSnapshot();
  });
  // expect(error('a=1}}}}{{{{')).toMatch('unbalanced-braces');
});

describe('CHARACTERS', () => {
  const ref = convertLatexToMarkup('J0');
  test.each([
    '^^4a0',
    '^^^^004a0',
    '\\char"4A 0',
    "\\char'0112 0",
    '\\char74 0',
    '\\char "004A 0',
    '\\char`J 0',
    '\\char`\\J 0',
    '\\char `\\J 0',
    '\\char   `\\J 0',
    '\\char +- +-  `\\J 0',
    '\\char +- -  `\\J 0',
    '\\char +- -- -++ `\\J 0',
    '\\unicode{"4A} 0',
    '\\unicode{"004A} 0',
    '\\unicode{x004A} 0',
  ])('%#/ %p renders as "J0"', (x) => {
    expect(convertLatexToMarkup(x)).toEqual(ref);
  });
});
describe('EXPANSION PRIMITIVES', () => {
  test.each([
    // ['\\obeyspaces =   =', '=\\space\\space\\space='],
    ['\\csname alpha\\endcsname', '\\alpha'],
    ['\\csname alph\\char"41\\endcsname', '\\alph A'],
    ['=\\sqrt\\bgroup x \\egroup=', '=\\sqrt{x}='],
    ['\\string\\alpha', '\\backslash alpha'],
    ['#?', '\\placeholder{}'],
  ])('%#/ %p matches %p', (a, b) => {
    expect(convertLatexToMarkup(a)).toMatch(convertLatexToMarkup(b));
  });
});

describe('ARGUMENTS', () => {
  test.each([
    ['a^\\frac12', 'a^{\\frac{1}{2}}'],
    ['\\sqrt3^2', '\\sqrt{3}^{2}'],
    ['\\frac12', '\\frac{1}{2}'],
    ['\\frac  1  2', '\\frac{1}{2}'],
    ['\\frac357', '\\frac{3}{5}7'],
    ['\\frac3a', '\\frac{3}{a}'],
    ['\\frac\\alpha\\beta', '\\frac{\\alpha}{\\beta}'],
    // ['\\frac{{1}}{2}', '\\frac{1}{2}'],
    ['\\frac  {  { 1  } } { 2 }', '\\frac{{1}}{2}'],
  ])('%#/ %p matches %p', (a, b) => {
    expect(convertLatexToMarkup(a)).toMatch(convertLatexToMarkup(b));
  });
  test.each(['\\frac', '\\frac{}', '\\frac{}{}'])(
    '%#/ %p renders correctly',
    (x) => {
      expect(markupAndError(x)).toMatchSnapshot();
    }
  );
});

describe('INFIX COMMANDS', () => {
  test.each([
    ['a\\over b', '\\frac{a}{b}'],
    ['a\\over b c', '\\frac{a}{bc}'],
    ['x{a+1\\over1-b}y', 'x{\\frac{a+1}{1-b}}y'],
    ['x{a+1\\over1-b\\over2}y', 'x{a+1\\over1-b2}y'],
  ])('%#/ %p matches %p', (a, b) => {
    expect(convertLatexToMarkup(a)).toMatch(convertLatexToMarkup(b));
  });

  expect(error('a\\over b \\over c')).toMatch('too-many-infix-commands');
});

describe('VARIANT SERIALIZATION (issue #2867)', () => {
  function serialize(latex: string): string {
    const atoms = parseLatex(latex, { parseMode: 'math' });
    return Atom.serialize(atoms, { defaultMode: 'math' });
  }

  test.each([
    // Single digit superscripts/subscripts don't have braces
    ['\\mathbb{R}^{0}', '\\mathbb{R}^0'],
    ['\\mathbb{R}^0', '\\mathbb{R}^0'],
    ['\\mathbb{N}_{1}', '\\mathbb{N}_1'],
    ['\\mathcal{F}^{2}', '\\mathcal{F}^2'],
    // Single letter subscripts/superscripts keep braces
    ['\\mathfrak{g}_{n}', '\\mathfrak{g}_{n}'],
    // Multi-character subscripts/superscripts should have braces
    ['\\mathbb{R}^{10}', '\\mathbb{R}^{10}'],
    ['\\mathbb{N}_{abc}', '\\mathbb{N}_{abc}'],
  ])('%#/ %p serializes as %p', (input, expected) => {
    expect(serialize(input)).toBe(expected);
  });
});
