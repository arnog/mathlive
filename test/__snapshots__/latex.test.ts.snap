// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ARGUMENTS 0/ "\\\\frac" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.5em"></span><span class="ML__strut--bottom" style="height:1.19em;vertical-align:-0.68em"></span><span class="ML__base"><span class="ML__mfrac"><span class="ML__nulldelimiter ML__open" style="width:0.12em"></span><span class="ML__vlist-t ML__vlist-t2"><span class="ML__vlist-r"><span class="ML__vlist" style="height:0.5em"><span class="ML__center" style="top:-2.31em"><span class="ML__pstrut" style="height:3em"></span><span style="height:0;display:inline-block"> </span></span><span style="top:-3.23em"><span class="ML__pstrut" style="height:3em"></span><span class="ML__frac-line" style="height:0.04em;display:inline-block"></span></span><span class="ML__center" style="top:-3.5em"><span class="ML__pstrut" style="height:3em"></span><span style="height:0;display:inline-block"> </span></span></span><span class="ML__vlist-s">​</span></span><span class="ML__vlist-r"><span class="ML__vlist" style="height:0.69em"></span></span></span><span class="ML__nulldelimiter ML__close" style="width:0.12em"></span></span></span></span>",
  "no-error",
]
`;

exports[`ARGUMENTS 1/ "\\\\frac{}" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.5em"></span><span class="ML__strut--bottom" style="height:1.19em;vertical-align:-0.68em"></span><span class="ML__base"><span class="ML__mfrac"><span class="ML__nulldelimiter ML__open" style="width:0.12em"></span><span class="ML__vlist-t ML__vlist-t2"><span class="ML__vlist-r"><span class="ML__vlist" style="height:0.5em"><span class="ML__center" style="top:-2.31em"><span class="ML__pstrut" style="height:3em"></span><span style="height:0;display:inline-block"> </span></span><span style="top:-3.23em"><span class="ML__pstrut" style="height:3em"></span><span class="ML__frac-line" style="height:0.04em;display:inline-block"></span></span><span class="ML__center" style="top:-3.5em"><span class="ML__pstrut" style="height:3em"></span><span style="height:0;display:inline-block"></span></span></span><span class="ML__vlist-s">​</span></span><span class="ML__vlist-r"><span class="ML__vlist" style="height:0.69em"></span></span></span><span class="ML__nulldelimiter ML__close" style="width:0.12em"></span></span></span></span>",
  "no-error",
]
`;

exports[`ARGUMENTS 2/ "\\\\frac{}{}" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.5em"></span><span class="ML__strut--bottom" style="height:1.19em;vertical-align:-0.68em"></span><span class="ML__base"><span class="ML__mfrac"><span class="ML__nulldelimiter ML__open" style="width:0.12em"></span><span class="ML__vlist-t ML__vlist-t2"><span class="ML__vlist-r"><span class="ML__vlist" style="height:0.5em"><span class="ML__center" style="top:-1.33em"><span class="ML__pstrut" style="height:2.02em"></span><span style="height:0;display:inline-block"></span></span><span style="top:-2.25em"><span class="ML__pstrut" style="height:2.02em"></span><span class="ML__frac-line" style="height:0.04em;display:inline-block"></span></span><span class="ML__center" style="top:-2.52em"><span class="ML__pstrut" style="height:2.02em"></span><span style="height:0;display:inline-block"></span></span></span><span class="ML__vlist-s">​</span></span><span class="ML__vlist-r"><span class="ML__vlist" style="height:0.69em"></span></span></span><span class="ML__nulldelimiter ML__close" style="width:0.12em"></span></span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 0/ "x" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.44em"></span><span class="ML__base"><span class="ML__mathit">x</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 1/ " x " renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.44em"></span><span class="ML__base"><span class="ML__mathit">x</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 2/ "%" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0"></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 3/ "% comment" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0"></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 4/ "x % comment" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.44em"></span><span class="ML__base"><span class="ML__mathit">x</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 5/ "x" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.44em"></span><span class="ML__base"><span class="ML__mathit">x</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 6/ "-12" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.65em"></span><span class="ML__strut--bottom" style="height:0.73em;vertical-align:-0.08em"></span><span class="ML__base"><span class="ML__cmr">−12</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 7/ "1234|/@.\`abcdefgzABCDEFGZ" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.75em"></span><span class="ML__strut--bottom" style="height:1em;vertical-align:-0.25em"></span><span class="ML__base"><span class="ML__cmr">1234∣/@.‘</span><span class="ML__mathit">abcde</span><span class="ML__mathit" style="margin-right:0.11em">f</span><span class="ML__mathit" style="margin-right:0.04em">g</span><span class="ML__mathit" style="margin-right:0.05em">z</span><span class="ML__mathit">A</span><span class="ML__mathit" style="margin-right:0.06em">B</span><span class="ML__mathit" style="margin-right:0.08em">C</span><span class="ML__mathit" style="margin-right:0.03em">D</span><span class="ML__mathit" style="margin-right:0.06em">E</span><span class="ML__mathit" style="margin-right:0.14em">F</span><span class="ML__mathit">G</span><span class="ML__mathit" style="margin-right:0.08em">Z</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 8/ "a b" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.7em"></span><span class="ML__base"><span class="ML__mathit">ab</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 9/ "ab" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.7em"></span><span class="ML__base"><span class="ML__mathit">ab</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 10/ "a~b" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.7em"></span><span class="ML__base"><span class="ML__mathit">a</span><span class="ML__cmr"> </span><span class="ML__mathit">b</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 11/ "a\\\\space b" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.7em"></span><span class="ML__base"><span class="ML__mathit">a</span><span class="ML__mediumspace"></span><span class="ML__mathit">b</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 12/ "{a}b" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.7em"></span><span class="ML__base"><span class="ML__mathit">a</span><span class="ML__mathit">b</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 13/ "{-}" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.59em"></span><span class="ML__strut--bottom" style="height:0.67em;vertical-align:-0.08em"></span><span class="ML__base"><span class="ML__cmr">−</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 14/ "-a" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.59em"></span><span class="ML__strut--bottom" style="height:0.67em;vertical-align:-0.08em"></span><span class="ML__base"><span class="ML__cmr">−</span><span class="ML__mathit">a</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 15/ "a-" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.59em"></span><span class="ML__strut--bottom" style="height:0.67em;vertical-align:-0.08em"></span><span class="ML__base"><span class="ML__mathit">a</span><span style="display:inline-block;width:0.23em"></span><span class="ML__cmr">−</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 16/ "a-b" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.7em"></span><span class="ML__strut--bottom" style="height:0.78em;vertical-align:-0.08em"></span><span class="ML__base"><span class="ML__mathit">a</span><span style="display:inline-block;width:0.23em"></span><span class="ML__cmr">−</span><span style="display:inline-block;width:0.23em"></span><span class="ML__mathit">b</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 17/ "a
b" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.7em"></span><span class="ML__base"><span class="ML__mathit">ab</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 18/ "a=1}" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.65em"></span><span class="ML__base"><span class="ML__mathit">a</span><span style="display:inline-block;width:0.28em"></span><span class="ML__cmr">=</span><span style="display:inline-block;width:0.28em"></span><span class="ML__cmr">1</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 19/ "a=1{" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.65em"></span><span class="ML__base"><span class="ML__mathit">a</span><span style="display:inline-block;width:0.28em"></span><span class="ML__cmr">=</span><span style="display:inline-block;width:0.28em"></span><span class="ML__cmr">1</span></span></span>",
  "unbalanced-braces",
]
`;

exports[`BASIC PARSING 20/ "a=1{}" renders correctly 1`] = `
[
  "<span class="ML__latex"><span class="ML__strut" style="height:0.65em"></span><span class="ML__base"><span class="ML__mathit">a</span><span style="display:inline-block;width:0.28em"></span><span class="ML__cmr">=</span><span style="display:inline-block;width:0.28em"></span><span class="ML__cmr">1</span></span></span>",
  "no-error",
]
`;
